build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -229,11 +248,40 @@ service ActionCache {
 // `INVALID_ARGUMENT` error will be returned. In either case, the client should
 // not attempt to retry the upload.
 //
-// For downloading blobs, the client must use the
-// [Read method][google.bytestream.ByteStream.Read] of the ByteStream API, with
-// a `resource_name` of `"{instance_name}/blobs/{hash}/{size}"`, where
-// `instance_name` is the instance name (see above), and `hash` and `size` are
-// the [Digest][build.bazel.remote.execution.v2.Digest] of the blob.
+// Small downloads can be grouped and requested in a batch via
+// [BatchReadBlobs][build.bazel.remote.execution.v2.ContentAddressableStorage.BatchReadBlobs].
+//
+// For large downloads, the client must use the
+// [Read method][google.bytestream.ByteStream.Read] of the ByteStream API.
```

Comment:
I wonder if we should be explicit that the read_offset, read_limit (and similarly, write_offset on WriteRequest) all refer to the uncompressed data? I don't see a way where it makes sense for it to refer to the compressed data, given the server may choose any compression level it wants, but it sounds useful to be explicit.

Reply:
Great questions.

First, some points to consider:
* Different implementations of the same compression algorithm can give different compressed data with the same compression level.
* There is not necessarily a direct mapping between a position in uncompressed data and a position in the compressed data.
* Servers can provide different compressed representations of the same data for any `Read`, but this must not change within a given `Read`.
* Servers want to store data in compressed form, and probably not also in uncompressed form.

Let's consider downloads first.
* The server can't cheaply find an arbitrary `read_offset` in the uncompressed data, because it's likely to be stored in compressed form. (This would require uncompressing some or all of the data, then re-compressing it from an offset.)
* The client can't start with an arbitrary `read_offset` in compressed data, because it does not know which compressed representation it will get.
* So clients must always start with a `read_offset` of `0`, and it should refer to the specific compressed representation of the data the server is currently providing, and must not be larger than the size of the stream sent by the server so far.
* Clients don't know the size of the compressed data ahead of time, to they must always specify `0` ("no limit").

Unfortunately, this means that clients cannot start a new `Read` to complete a previous interrupted/incomplete `Read`.

Now consider uploads, which are a bit simpler:
* The `write_offset` should refer to the compressed data representation that the client is sending.
* The `write_offset` should either always begin at `0`, **or perhaps** servers could choose whether or not to allow resumed uploads based on the uniqueness of the `guid` and `uncompressed_hash` pair.

Reply:
@EricBurnett: my reasoning above differs from yours IIRC, for downloads in particular. It mostly hinges on how difficult it is to find a particular offset in the uncompressed data given only a compressed version of it.

In the worst case scenario I suspect this might devolve into decompressing and recompressing almost the entire blob, in which case you could instead use gRPC level compression with the uncompressed bytestream resource.

Reply:
If I follow your reasoning correctly, it's conditioned on the concern:
> The server can't cheaply find an arbitrary read_offset in the uncompressed data, because it's likely to be stored in compressed form.

This is not quite true.

- A basic server can start simply as you suggest, and only support reads and writes from offset 0. Totally reasonable imho.
- One step up from that, a server can seek and discard server-side for offset reads (better than sending it over the wire unnecessarily), and can append new chunks to the end of the compressed contents for resumable writes.
  - Reads are not super efficient, but semantically appropriate, and gracefully transition to more complicated implementations if usage is high enough to require the efficiency.
  - Resuming broken writes requires recompressing from offset N. (Note that the server must close the compressor and write it out cleanly for the initial partial upload - appending to a truncated stream would not work). Also works for e.g. LogStream - each upload is an independently compressed chunk of bytes, with no knowledge of the compression of the ones that came before required.
- Finally, servers can store data compressed in chunks, to permit efficient offset reads also. (Decompress nearest chunk and recompress desired portion of it).
   - Of note here, for most compressors (including zstd and gzip) the output is concatenatable - `decompress(compress("foo")+compress("bar"))=="foobar"`. So servers can instead of storing `compress(file)`, store `[compress(file.chunk1), compress(file.chunk2), ...]` . And then to return the whole stream to the user, return the concatenation of all these chunks; to return contents from a later offset, recompress the one partial chunk (`compress(decompress(file.chunkK)[N:])`) and return that followed by the concatenation of later chunks in sequence.

Compare with the semantics of offsets expressed in compressed bytes:
 - Arbitrary offset reads are meaningless - clients will not be able to decode from byte N without knowing the state of the bytes preceding.
 - Offset reads *to resume a broken stream* are possible, assuming the server guarantees to return exactly the same byte sequence. Without that guarantee, not usable.
 - Offset writes are similarly useful *to resume a broken stream*. Observing a broken write and re-compressing bytes starting at offset N wouldn't work; the exact compressor output must be kept, so it must be literal resumption of the exact output.
 - This is all considerably harder if the server supports multiple compression schemes and recompresses on the fly for any of them - it would have to be able to reproduce state without simply writing to or reading from the compressed bytes as provided by the user.

Overall, we can see offsets into compressed bytes as more of a "resumption token" than anything else - use of them doesn't permit any other use-cases.

To me, then, offsets into the uncompressed bytes are much better: they can be used to permit efficient and semantically-meaningful offset reads, broken and multi-party writes, and work well with recompression between compressors. They're slightly harder to use for resuming broken writes, but in all other respects seem like a clear win to me.

Reply:
> If I follow your reasoning correctly, it's conditioned on the concern:
> 
> > The server can't cheaply find an arbitrary read_offset in the uncompressed data, because it's likely to be stored in compressed form.
> 
> This is not quite true.
> 
> * A basic server can start simply as you suggest, and only support reads and writes from offset 0. Totally reasonable imho.

This is reasonable, if there's a clear way to tell the client "seeking not supported". OUT_OF_RANGE kind of makes sense, but technically the request is in range but unsupported. OUT_OF_RANGE feels even less accurate for discontiguous ReadRequests.

If we can find a clean solution to this, then I'm fine with offsets referring to uncompressed data.

> * Finally, servers can store data compressed in chunks, to permit efficient offset reads also. (Decompress nearest chunk and recompress desired portion of it).
>   
>   * Of note here, for most compressors (including zstd and gzip) the output is concatenatable - `decompress(compress("foo")+compress("bar"))=="foobar"`. So servers can instead of storing `compress(file)`, store `[compress(file.chunk1), compress(file.chunk2), ...]` . And then to return the whole stream to the user, return the concatenation of all these chunks; to return contents from a later offset, recompress the one partial chunk (`compress(decompress(file.chunkK)[N:])`) and return that followed by the concatenation of later chunks in sequence.

One annoying part here is that (I suspect) splitting compressed data into concatenatable chunks is probably difficult. In practice this probably means that the server would need to recompress the blob before storing it in chunked form. Though I supposed it would need to be done at most once per blob, and an optional implementation detail.

> * This is all considerably harder if the server supports multiple compression schemes and recompresses on the fly for any of them - it would have to be able to reproduce state without simply writing to or reading from the compressed bytes as provided by the user.

Right, but isn't the goal to minimise server side recompression?

> Overall, we can see offsets into compressed bytes as more of a "resumption token" than anything else - use of them doesn't permit any other use-cases.
> 
> To me, then, offsets into the uncompressed bytes are much better: they can be used to permit efficient and semantically-meaningful offset reads, broken and multi-party writes, and work well with recompression between compressors. They're slightly harder to use for resuming broken writes, but in all other respects seem like a clear win to me.

I agree re semantic meaningfulness, and to a lesser extent re efficiency (still OK with me if servers can tell the clients to just start over).

Reply:
> This is reasonable, if there's a clear way to tell the client "seeking not supported".

Good point! Yes, we should definitely clarify that if it's to be allowed. Though that said, that also puts the burden on clients to have a codepath for falling back if an offset read is rejected, and would be considered a backwards-incompatible API change (since clients presumably don't handle this today). The more I think about it, the more I realize it's probably better to require servers at least implement an inefficient read-and-discard path. That's arguably simpler on net (the client-side fallback would be uglier and embed read-and-discard within it), and in most cases it'll be rarely enough hit that inefficient is probably fine?

> One annoying part here is that (I suspect) splitting compressed data into concatenatable chunks is probably difficult. In practice this probably means that the server would need to recompress the blob before storing it in chunked form.

Yeah, it means the server needs to recompress, but that's just a combination of two things it has to do anyways - for compressed uploads it must be at least _decompressing_ so it can check the digest matches; for uncompressed uploads it will be compressing as it puts the file into storage; now it needs to chain both. And yeah, chunking is more complicated if you choose to implement that, but it's not _too_ bad. (I've already written it for our own compressed storage implementation fwiw - didn't take me too long).

> Though I supposed it would need to be done at most once per blob

And for most use cases I'd expect reads to outweigh writes by a decent margin (we observe 9:1), so in practice only a small fraction of the bytes being handled by the server require compression. I found it to have a negligible impact on net server CPU usage, when we switched from uncompressed storage to chunk-compressed zstd-1. YMMV.

> > This is all considerably harder if the server supports multiple compression schemes and recompresses on the fly for any of them - it would have to be able to reproduce state without simply writing to or reading from the compressed bytes as provided by the user.

> Right, but isn't the goal to minimise server side recompression?

I realized my example was over-complicated - we don't even need to bring other compressors into it. Any uncompressed read of a blob that's stored compressed is exactly the same - a read with an optional offset into the uncompressed contents - which is something everyone has to implement anyways.

Reply:
> > This is reasonable, if there's a clear way to tell the client "seeking not supported".
> 
> Good point! Yes, we should definitely clarify that if it's to be allowed. Though that said, that also puts the burden on clients to have a codepath for falling back if an offset read is rejected, and would be considered a backwards-incompatible API change (since clients presumably don't handle this today). The more I think about it, the more I realize it's probably better to require servers at least implement an inefficient read-and-discard path. That's arguably simpler on net (the client-side fallback would be uglier and embed read-and-discard within it), and in most cases it'll be rarely enough hit that inefficient is probably fine?

I guess we should make an explicit statement here, one of:
1) Servers MUST support compressed reads from arbitrary offsets.
2) Servers MAY return error code X for non-zero offset reads, if they are not supported. Maybe `ABORTED` makes sense, based on the discussion in https://developers.google.com/maps-booking/reference/grpc-api/status_codes#failed_precondition_vs_aborted_vs_unavailable

> > One annoying part here is that (I suspect) splitting compressed data into concatenatable chunks is probably difficult. In practice this probably means that the server would need to recompress the blob before storing it in chunked form.
> 
> Yeah, it means the server needs to recompress, but that's just a combination of two things it has to do anyways - for compressed uploads it must be at least _decompressing_ so it can check the digest matches; for uncompressed uploads it will be compressing as it puts the file into storage; now it needs to chain both. And yeah, chunking is more complicated if you choose to implement that, but it's not _too_ bad. (I've already written it for our own compressed storage implementation fwiw - didn't take me too long).
> 
> > Though I supposed it would need to be done at most once per blob
> 
> And for most use cases I'd expect reads to outweigh writes by a decent margin (we observe 9:1), so in practice only a small fraction of the bytes being handled by the server require compression. I found it to have a negligible impact on net server CPU usage, when we switched from uncompressed storage to chunk-compressed zstd-1. YMMV.

After some experimentation this seems OK (though I haven't benchmarked at scale yet). I'll try to update the PR tomorrow.

Reply:
> > > This is reasonable, if there's a clear way to tell the client "seeking not supported".
> > 
> > 
> > Good point! Yes, we should definitely clarify that if it's to be allowed. Though that said, that also puts the burden on clients to have a codepath for falling back if an offset read is rejected, and would be considered a backwards-incompatible API change (since clients presumably don't handle this today). The more I think about it, the more I realize it's probably better to require servers at least implement an inefficient read-and-discard path. That's arguably simpler on net (the client-side fallback would be uglier and embed read-and-discard within it), and in most cases it'll be rarely enough hit that inefficient is probably fine?
> 
> I guess we should make an explicit statement here, one of:
> 
> 1. Servers MUST support compressed reads from arbitrary offsets.
> 2. Servers MAY return error code X for non-zero offset reads, if they are not supported. Maybe `ABORTED` makes sense, based on the discussion in https://developers.google.com/maps-booking/reference/grpc-api/status_codes#failed_precondition_vs_aborted_vs_unavailable

1. Should we add support for offset reads to the capabilities? That would make choosing the client side codepath dead easy, and would save servers from needing to implement offset reads. I'm leaning against it--the throw-bytes-away approach is simple enough--but wondering how others feel.

2. If we make not supporting (non-zero) offset reads an option, then I think that INVALID_INPUT is the correct error code. The non-zero offset is not a supported input.
> 
> > > One annoying part here is that (I suspect) splitting compressed data into concatenatable chunks is probably difficult. In practice this probably means that the server would need to recompress the blob before storing it in chunked form.
> > 
> > 
> > Yeah, it means the server needs to recompress, but that's just a combination of two things it has to do anyways - for compressed uploads it must be at least _decompressing_ so it can check the digest matches; for uncompressed uploads it will be compressing as it puts the file into storage; now it needs to chain both. And yeah, chunking is more complicated if you choose to implement that, but it's not _too_ bad. (I've already written it for our own compressed storage implementation fwiw - didn't take me too long).
> > > Though I supposed it would need to be done at most once per blob
> > 
> > 
> > And for most use cases I'd expect reads to outweigh writes by a decent margin (we observe 9:1), so in practice only a small fraction of the bytes being handled by the server require compression. I found it to have a negligible impact on net server CPU usage, when we switched from uncompressed storage to chunk-compressed zstd-1. YMMV.
> 
> After some experimentation this seems OK (though I haven't benchmarked at scale yet). I'll try to update the PR tomorrow.

Reply:
> 1. Should we add support for offset reads to the capabilities? That would make choosing the client side codepath dead easy, and would save servers from needing to implement offset reads. I'm leaning against it--the throw-bytes-away approach is simple enough--but wondering how others feel.
> 2. If we make not supporting (non-zero) offset reads an option, then I think that INVALID_INPUT is the correct error code. The non-zero offset is not a supported input.

I don't think it's worth adding a compressed-offset-reads boolean to the capabilities.

If we allow servers to return INVALID_INPUT for positive offset compressed reads, clients will have to do some guesswork to figure out which field was considered invalid. Whereas I think throw-the-bytes-away for compressed offset reads is simple enough for servers to implement, and uncommon enough to not spend time worrying about the performance impact.

Should we add some text to state that servers must support compressed reads from positive (in range) offsets?